### PR TITLE
DW-277: Added SF1601

### PR DESCRIPTION
**Pull request by rimi-itk**
https://jira.itkdev.dk/browse/DW-277

### CODE REVIEW

File: src/Service/SF1601/SF1601.php [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+                'body' => $document->saveXML(),
+                'transactionId' => $transactionId,
+            ]);
+
+        return $response;
+    }
+
+    private function buildKombiRequest(string $type, Message $message)
+    {
+        if (!in_array($type, [self::TYPE_DIGITAL_POST])) {
```
Comment (jekuaitk) [inline src/Service/SF1601/SF1601.php]: Is this made as preparation for other types? If not it seems unnecessary to use arrays for this.

File: src/Service/AbstractRESTService.php [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+    {
+        $resolver
+            ->setRequired([
+                'client_cert_pub',
+                'client_cert_key',
+                'authority_cvr',
+            ])
+            ->setDefaults([
+                'debug' => false,
+                'test_mode' => true,
```
Comment (jekuaitk) [inline src/Service/AbstractRESTService.php:314]: 👍

File: src/Service/AbstractRESTService.php [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+
+        return $this->request(
+            $method,
+            $url,
+            [
+                'headers' => [
+                        'Authorization' => sprintf('%s %s', $accessToken['token_type'], $accessToken['access_token']),
+                        'X-TransaktionsId' => $transactionId,
+                        'X-TransaktionsTid' => $transactionTid,
+                    ] + $headers,
```
Comment (jekuaitk) [inline src/Service/AbstractRESTService.php]: Im guessing documentation for this is hidden within nested zip files?

**Review by jekuaitk — APPROVED** [review]

**Review by jekuaitk — APPROVED** [review]

File: src/Service/AbstractRESTService.php [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+
+    private function getCache()
+    {
+        return $this->options['cache'];
+    }
+
+    private function getCacheKey(string $key, array $payload): string
+    {
+        return preg_replace(
+            '#[{}()/\\\\@:]+#',
```
Comment (jekuaitk) [inline src/Service/AbstractRESTService.php:183]: Duplicate match on `/` that could be removed.

File: src/Service/AbstractRESTService.php [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+     */
+    private function getSAMLToken(): string
+    {
+        $cache = $this->getCache();
+        $cacheKey = $this->getCacheKey(__METHOD__, [
+            'AnvenderKontekst' => ['Cvr' => $this->options['authority_cvr']],
+            'AppliesTo' => ['EndpointReference' => ['Address' => $this->options['svc_entity_id']]],
+        ]);
+
+        $expirationTimeOffset = '-15 minutes';
```
Comment (jekuaitk) [inline src/Service/AbstractRESTService.php]: If i remember correctly these can be used for 8 hours so 15 minutes seems reasonable, but should this be configurable?

File: src/Service/AbstractRESTService.php [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+    private function getAccessToken(string $samlToken): array
+    {
+        $cache = $this->getCache();
+        $cacheKey = $this->getCacheKey(__METHOD__, [
+            'access_token_svc' => $this->options['access_token_svc'],
+            'saml_token' => $samlToken,
+        ]);
+
+        // Cache expire offset in seconds (cf. ItemInterface::expiresAfter).
+        $expiresAfterOffset = 60;
```
Comment (jekuaitk) [inline src/Service/AbstractRESTService.php]: How long do these last? Im guessing that had an influence on why you chose 60.

File: src/Service/AbstractRESTService.php [hunk 7]
```diff
@@ -0,0 +1,338 @@
...
+        file_put_contents($certificateFilename, $this->getCertificate());
+
+        $privateKeyFilename = tempnam(sys_get_temp_dir(), 'cert');
+        file_put_contents($privateKeyFilename, $this->getPrivateKey());
+
+        return $this->client()->request($method, $url, $options + [
+                'local_cert' => $certificateFilename,
+                'local_pk' => $privateKeyFilename,
+                'on_progress' => function (int $dlNow, int $dlSize, array $info) use (&$certificateFilename, &$privateKeyFilename): void {
+                    // Delete temporary certificate files when receiving response headers.
```
Comment (jekuaitk) [inline src/Service/AbstractRESTService.php:124]: Are temporary files not temporary? Is this necessary?

**Review by jekuaitk — APPROVED** [review]
A few comments and questions, but looks fine.

File: src/Service/AbstractRESTService.php (hunk 7 above)
Comment (rimi-itk) [reply src/Service/AbstractRESTService.php:124]: No, temporary files are not temporary and must be deleted manually. I don't understand why this is the case. Maybe I don't understand what “temporary” means …

File: src/Service/AbstractRESTService.php (hunk 4 above)
Comment (rimi-itk) [reply src/Service/AbstractRESTService.php:183]: Do you mean `\`?

File: src/Service/AbstractRESTService.php (hunk 5 above)
Comment (rimi-itk) [reply src/Service/AbstractRESTService.php]: Option for this added in https://github.com/itk-dev/serviceplatformen/pull/9/commits/d1093f70bd08309507e53cf0d1bb88c448927f21.

File: src/Service/AbstractRESTService.php (hunk 6 above)
Comment (rimi-itk) [reply src/Service/AbstractRESTService.php]: Option for this added in https://github.com/itk-dev/serviceplatformen/pull/9/commits/d1093f70bd08309507e53cf0d1bb88c448927f21.

File: src/Service/AbstractRESTService.php (hunk 4 above)
Comment (jekuaitk) [reply src/Service/AbstractRESTService.php:183]: Yes!

Comment (rimi-itk) [reply src/Service/AbstractRESTService.php:183]: We need 4 backslashes in the code, 2 for the actual backslash and 2 for the backslash escaping the backslash. It's weird, but it's true. 3 backslashes are actually enough, but that looks weird.

See https://stackoverflow.com/questions/11044136/right-way-to-escape-backslash-in-php-regex and https://www.developwebsites.net/match-backslash-preg_match-php/ for some gory details.

**Review by jekuaitk — APPROVED** [review]

**Review by jekuaitk — APPROVED** [review]

**Pull request merged**